core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java

```diff
@@ -604,39 +579,62 @@ public int register(String subject,
         }
       }
 
+      Config config = getConfigInScope(subject);
       Mode mode = getModeInScope(subject);
-      boolean isCompatible = true;
-      List<String> compatibilityErrorLogs = new ArrayList<>();
+
+      boolean modifiedSchema = false;
       if (mode != Mode.IMPORT) {
-        Config config = getConfigInScope(subject);
-        if (schemaId < 0) {
-          parsedSchema = maybePopulateFromPrevious(config, schema, parsedSchema, undeletedVersions);
+        modifiedSchema = maybePopulateFromPrevious(config, schema, undeletedVersions);
+      }
+
+      int schemaId = schema.getId();
+      ParsedSchema parsedSchema = canonicalizeSchema(schema, schemaId < 0, normalize);
+
+      if (parsedSchema != null) {
+        // see if the schema to be registered already exists
+        SchemaIdAndSubjects schemaIdAndSubjects = this.lookupCache.schemaIdAndSubjects(schema);
+        if (schemaIdAndSubjects != null
+            && (schemaId < 0 || schemaId == schemaIdAndSubjects.getSchemaId())) {
+          if (schemaIdAndSubjects.hasSubject(subject)
+              && !isSubjectVersionDeleted(subject, schemaIdAndSubjects.getVersion(subject))) {
+            // return only if the schema was previously registered under the input subject
+            return modifiedSchema
+                ? schema.copy(
+                    schemaIdAndSubjects.getVersion(subject), schemaIdAndSubjects.getSchemaId())
+                : new Schema(subject, schemaIdAndSubjects.getSchemaId());
```

Comment:
What does this mean? If we re-register the same schema, but the schema needs to populate the ruleset/metadata, will it still be considered as same, or should it be a new schema?

Reply:
This block returns if the schema already exists.  The `modifiedSchema` indicates whether the metadata/ruleSet was populated.  So a schema can already exist even after populating the metadata/ruleSet.